sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteNonAggregateFirst.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, Literal}
+import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, First}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.FIRST
+import org.apache.spark.sql.types.IntegerType
+
+/**
+ * Rewrite aggregate plan with a single [[First]] function when grouping is absent. In such a case
+ * the query is equivalent to simple projection with limit 1.
+ *
+ * Input Pseudo-Query:
+ * {{{
+ *   SELECT FIRST(col) FROM table
+ * }}}
+ *
+ * Rewritten Query:
+ * {{{
+ *   SELECT col FROM table LIMIT 1
+ * }}}
+ *
+ * Note that using IGNORE NULLS with [[First]] blocks rewrite logic since projection with NOT NULL
+ * filter might return different result than [[First]] if all values are NULL.
+ */
+object RewriteNonAggregateFirst extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan.transformWithPruning(_.containsPattern(FIRST), ruleId) {
+      case agg: Aggregate if isNonAggregateFirst(agg) =>
+        rewriteNonAggregateFirstToProject(agg)
+    }
+  }
+
+  private def isNonAggregateFirst(agg: Aggregate): Boolean = {
+    agg.aggregateExpressions.length == 1 && agg.groupingExpressions.isEmpty
+  }
+
+  private def rewriteNonAggregateFirstToProject(agg: Aggregate): LogicalPlan = {
+    var skipRewrite = false
+    val projectList = agg.aggregateExpressions.map {
+      case alias: Alias =>
```

Comment:
I think we could simplify this by just matching against the Alias <- First sequence in one go, e.g. `case a@Alias(f: First, _) if !f.ignoreNulls => Some(a.copy(child = f.child))` and then catch all other cases with `case _ => None`. Then at the end you can check `if (projectList.forall(_.isDefined) { ... projectList.map(_.get) } else { agg }`.

Reply:
redid since the logic changed.